File: tests/test_mpi3rma_fortran/dynamic_ddt_5.f90
      subroutine test_mpi3rma_dynamic_ddt_5( base, array_size, win, comm )

      use utilities, only: check_fompi_status

#if INCLUDE_TYPE==1
      implicit none
      
      include 'fompif.h'
#endif
#if INCLUDE_TYPE==2
      use fompi
      
      implicit none
#endif

      integer, intent(in) :: array_size
      integer, dimension(array_size), intent(inout) :: base
      integer, intent(in) :: win
      integer, intent(in) :: comm

      integer :: i
      integer :: k
      integer :: rank
      integer :: commsize
      integer :: mpierr
      integer(kind=MPI_ADDRESS_KIND) :: start

      integer :: target_dtype
      integer :: origin_dtype
      integer, dimension(3) :: blocklength
      integer, dimension(3) :: displacement

      character(50) :: filename

      write (filename,*) "dynamic_ddt_5.f90"

      call MPI_Comm_rank( comm, rank, mpierr )
      call MPI_Comm_size( comm, commsize, mpierr )

      if (rank == 0) then
        do i=4,6
          base(i) = 0;
        enddo
        base(9) = 0
        do i=12,14
          base(i) = 0
        enddo
      endif
      do i=0,2 
        base(i+1) = rank+i
      enddo
      do i=0,1
        base(7+i) = rank+i+3
      enddo
      do i=0,1
        base(10+i) = rank+i+5
      enddo
 
      blocklength(1) = 3
      blocklength(2) = 1
      blocklength(3) = 3
      displacement(1) = 3
      displacement(2) = 8
      displacement(3) = 11

      call MPI_Type_indexed( 3, blocklength(1), displacement(1), MPI_INTEGER, target_dtype, mpierr )
      call MPI_Type_commit( target_dtype, mpierr )

      blocklength(1) = 3
      blocklength(2) = 2
      blocklength(3) = 2
      displacement(1) = 0
      displacement(2) = 6
      displacement(3) = 9

      call MPI_Type_indexed( 3, blocklength(1), displacement(1), MPI_INTEGER, origin_dtype, mpierr )
      call MPI_Type_commit( origin_dtype, mpierr )

      call foMPI_Get_address( base(1), start, mpierr )
      call MPI_Bcast( start, 1, MPI_AINT, 0, comm, mpierr )

      call foMPI_Win_fence( foMPI_MODE_NOPRECEDE, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 77, rank)

      call foMPI_Accumulate( base(1), 1, origin_dtype, 0, start, 1, target_dtype, foMPI_SUM, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 80, rank)
 
      call foMPI_Win_fence( foMPI_MODE_NOSUCCEED, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 83, rank)

      call foMPI_Type_free( target_dtype, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 86, rank)
      call foMPI_Type_free( origin_dtype, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 88, rank)

      if ( rank .eq. 0 ) then
        do k=0,2
          if( base(4+k) .ne. ((commsize-1)*commsize/2+k*commsize) ) then
            write (*,*) "Error on rank ", rank, " in ", trim(filename), &
              " line ", 92, ": element ", 4+k, " is ", base(4+k), &
              " instead of ", (commsize-1)*commsize/2+k*commsize, "."
            STOP
          endif
        enddo
        if( base(9) .ne. ((commsize-1)*commsize/2+3*commsize) ) then
          write (*,*) "Error on rank ", rank, " in ", trim(filename), &
            " line ", 99, ": element ", 49, " is ", base(9), &
            " instead of ", (commsize-1)*commsize/2+3*commsize, "."
          STOP
        endif
        do k=0,2
          if( base(12+k) .ne. ((commsize-1)*commsize/2+(4+k)*commsize) ) then
            write (*,*) "Error on rank ", rank, " in ", trim(filename), &
              " line ", 106, ": element ", 12+k, " is ", base(12+k), &
              " instead of ", (commsize-1)*commsize/2+(4+k)*commsize, &
              "."
            STOP
          endif
        enddo
      endif

      if ( rank .eq. 0 ) then
        write (*,*) "No Error"
      endif

     end subroutine test_mpi3rma_dynamic_ddt_5

File: tests/test_foMPI/dynamic_ddt_5.c
#include "fompi.h"
#include "test_mpi3rma.h"

void test_mpi3rma_dynamic_ddt_5( int* base, foMPI_Win win, MPI_Comm comm ) {

  int i;
  int k;
  int rank;
  int commsize;
  int status;
  MPI_Aint start;

  MPI_Datatype target_dtype;
  MPI_Datatype origin_dtype;
  int blocklength[3];
  int displacement[3];

  MPI_Comm_rank( comm, &rank );
  MPI_Comm_size( comm, &commsize );

  if (rank == 0) {
    for( i=3 ; i<6 ; i++ ) {
      base[i] = 0;
    }
    base[8] = 0;
    for( i=11 ; i<14 ; i++ ) {
      base[i] = 0;
    }
  }
  for( i=0 ; i<3 ; i++ ) {
    base[i] = rank+i;
  }
  for( i=0 ; i<2 ; i++ ) {
    base[6+i] = rank+i+3;
  }
  for( i=0 ; i<2 ; i++ ) {
    base[9+i] = rank+i+5;
  }
 
  blocklength[0] = 3;
  blocklength[1] = 1;
  blocklength[2] = 3;
  displacement[0] = 3;
  displacement[1] = 8;
  displacement[2] = 11;

  MPI_Type_indexed(3, &blocklength[0], &displacement[0], MPI_INT, &target_dtype);
  MPI_Type_commit(&target_dtype);

  blocklength[0] = 3;
  blocklength[1] = 2;
  blocklength[2] = 2;
  displacement[0] = 0;
  displacement[1] = 6;
  displacement[2] = 9;

  MPI_Type_indexed(3, &blocklength[0], &displacement[0], MPI_INT, &origin_dtype);
  MPI_Type_commit(&origin_dtype);

  MPI_Get_address( &base[0], &start );
  MPI_Bcast( &start, 1, MPI_AINT, 0, comm );

  status = foMPI_Win_fence( foMPI_MODE_NOPRECEDE, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  status = foMPI_Accumulate( &base[0], 1, origin_dtype, 0, start, 1, target_dtype, foMPI_SUM, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
 
  status = foMPI_Win_fence( foMPI_MODE_NOSUCCEED, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  foMPI_Type_free( &target_dtype );
  foMPI_Type_free( &origin_dtype );

  if ( rank == 0 ) {
    for( k=0 ; k<3 ; k++ ) {
      if( base[3+k] != ((commsize-1)*commsize/2+k*commsize) ) {
        fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, 3+k, base[3+k], (commsize-1)*commsize/2+k*commsize );
        exit(EXIT_FAILURE);
      }
    }
    if( base[8] != ((commsize-1)*commsize/2+3*commsize) ) {
      fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, 8, base[8], (commsize-1)*commsize/2+3*commsize );
      exit(EXIT_FAILURE);
    }
    for( k=0 ; k<3 ; k++ ) {
      if( base[11+k] != ((commsize-1)*commsize/2+(4+k)*commsize) ) {
        fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, 11+k, base[11+k], (commsize-1)*commsize/2+(4+k)*commsize );
        exit(EXIT_FAILURE);
      }
    }
  }

  if ( rank == 0 ) {
    printf("No Error\n");
  }

  return;
}

